src/WebJobs.Script/Host/ScriptHost.cs

```diff
@@ -1069,6 +1075,46 @@ internal static void ApplyConfiguration(JObject config, ScriptHostConfiguration
             }
         }
 
+        internal static string GetDefaultHostId(ScriptSettingsManager settingsManager, ScriptHostConfiguration scriptConfig)
+        {
+            // We're setting the default here on the newly created configuration
+            // If the user has explicitly set the HostID via host.json, it will overwrite
+            // what we set here
+            string hostId = null;
+            if (scriptConfig.IsSelfHost)
+            {
+                // When running locally, derive a stable host ID from machine name
+                // and root path. We use a hash rather than the path itself to ensure
+                // IDs differ (due to truncation) between folders that may share the same
+                // root path prefix.
+                // Note that such an ID won't work in distributed scenarios, so should
+                // only be used for local/CLI scenarios.
+                var sanitizedPath = scriptConfig.RootScriptPath
+                    .Where(char.IsLetterOrDigit)
+                    .Aggregate(new StringBuilder(), (b, c) => b.Append(c))
+                    .ToString();
+
+                hostId = $"{Environment.MachineName}-{Math.Abs(sanitizedPath.GetHashCode())}";
+            }
+            else if (!string.IsNullOrEmpty(settingsManager.AzureWebsiteDefaultSubdomain))
+            {
+                // If running on Azure Web App, derive the host ID from the default subdomain
+                // Trim any trailing - as they can cause problems with queue names
+                hostId = settingsManager.AzureWebsiteDefaultSubdomain.TrimEnd('-');
+            }
+
+            if (!string.IsNullOrEmpty(hostId))
+            {
+                if (hostId.Length > ScriptConstants.MaximumHostIdLength)
+                {
+                    // Truncate to the max host name length if needed
+                    hostId = hostId.Substring(0, ScriptConstants.MaximumHostIdLength);
+                }
```

Comment:
SDK validation is here: https://github.com/Azure/azure-webjobs-sdk/blob/663a508e8a851629c26a51e7de3af36629dfd120/src/Microsoft.Azure.WebJobs.Host/Executors/HostIdValidator.cs#L20

I think we're good on length.